### PR TITLE
Matthews Correlation Fix & Binary helpers

### PR DESCRIPTION
**Pull request by kuza55**
I added some extra functions to the backend for logical operations.

Originally I had these just based on existing backend functions (as in the Theano implementation) and thought they could just be helpers in common or something, but it seems Tensorflow has built-in operations for this already.

While rewriting the matthews correlation metric with this I noticed it was buggy, so I fixed that and added tests for the metric and the new backend functions.

I haven't benchmarked the Theano backend to see if this is an ideal implementation, but it at the very least seems consistent with tensorflow, so at least they're accurate.

[EDIT]: Actually, the backend tests are failing on my machine since it seems like the CTC bits are incompatible with TF 0.10.0rc0 so I couldn't check if the backend test works, so hopefully CI can verify this. Metric test passes though.

### CODE REVIEW

**Comment by kuza55** [comment]
Ok, so the TensorFlow backend ops operate on boolean tensors, which isn't something Theano doesn't even seem to support.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1830,11 +1850,11 @@ def ctc_decode(y_pred, input_length, greedy=True, beam_width=100,
     input_length = tf.to_int32(input_length)
 
     if greedy:
-        (decoded, log_prob) = ctc.ctc_greedy_decoder(
+        (decoded, log_prob) = ctc.ctc_greedy_decoder(
             inputs=y_pred,
             sequence_length=input_length)
     else:
-        (decoded, log_prob) = ctc.ctc_beam_search_decoder(
+        (decoded, log_prob) = ctc.ctc_beam_search_decoder(
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Whitespace changes?

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -1,9 +1,9 @@
 import tensorflow as tf
 from tensorflow.python.training import moving_averages
 try:
-    import tensorflow.contrib.ctc as ctc
+    import tensorflow.contrib.ctc as ctc
 except ImportError:
-    from tensorflow.python.ops import ctc_ops as ctc
+    from tensorflow.python.ops import ctc_ops as ctc
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Whitespace changes?

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -304,6 +304,26 @@ def all(x, axis=None, keepdims=False):
...
+def logical_and(x, y):
+    return clip(x + y - 1, 0, 1)
+
+
+def logical_or(x, y):
+    return clip(x + y, 0, 1)
+
+
+def logical_xor(x, y):
+    return 1 - abs(x + y - 1)
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: I am not seeing much reusability for these 5 functions at this point. Maybe it would be better to only change the Matthews correlation implementation without adding new functions to the backend.

**Comment by kuza55** [comment]
I'm not sure what happened to the whitespace since I didn't touch those lines, but I guess I can go in with a hex editor.

So, I'm also not the biggest fan of adding them to the back ends in this manner myself, if the back ends were classes I would add these to the parent class so that they could be shared. I can look at putting them in common again, but that also didn't feel right, maybe a backend_helpers.py?

However I think these are useful when you are defining your own control structures.

A question came up on the gitter chat that boiled down to doing an update with a logical or where the asker was originally looking for a way to do updates of a single index in a tensor, which is a much trickier proposition.

I had some need for these operations when writing my differentiable random forest layer, though I mostly needed not/and which are the more straight forward ones.

I would argue that the fact that these operations exist in TF on bool tensors is an indication that there is a desire to do this sort of thing.

But I mostly thought adding them would help people who are not used to expressing logic in this form see that it's possible and help with the trickier ones. There's all sorts of these kinda of tricks scattered around the codebase that I think would help people if they were documented and they could also be optimized in one place. Another thing in this line that I used recently was - (p - abs(p) / 2 to extract the negative component of a tensor, which I'd seen reading some PRs that I would have probably been scratching my head for a bit coming up with an alternative.

So a proper version would require some documentation to reach this goal, but lmk what you think.
